File: cloud/mdb/cli/dbaas/internal/metadb/cloud.py
from click import ClickException

from cloud.mdb.cli.dbaas.internal.db import db_query


def get_cloud(ctx, cloud_id):
    query = """
        SELECT
            cloud_ext_id "id",
            clusters_used,
            clusters_quota,
            cpu_used,
            cpu_quota,
            gpu_used,
            gpu_quota,
            memory_used,
            memory_quota,
            hdd_space_used,
            hdd_space_quota,
            ssd_space_used,
            ssd_space_quota
        FROM dbaas.clouds cl
        WHERE cloud_ext_id = %(cloud_id)s
        """
    result = db_query(ctx, 'metadb', query, cloud_id=cloud_id)

    if not result:
        raise ClickException(f'Cloud "{cloud_id}" not found.')

    return result[0]


def get_clouds(ctx, cluster_ids=None, limit=None):
    query = """
        SELECT
            cloud_ext_id "id",
            clusters_used,
            clusters_quota,
            cpu_used,
            cpu_quota,
            gpu_used,
            gpu_quota,
            memory_used,
            memory_quota,
            hdd_space_used,
            hdd_space_quota,
            ssd_space_used,
            ssd_space_quota
        FROM dbaas.clouds cl
        {% if cluster_ids %}
        WHERE EXISTS (SELECT 1
                      FROM dbaas.clusters
                      JOIN dbaas.folders USING (folder_id)
                      WHERE cloud_id = cl.cloud_id AND cid = ANY(%(cluster_ids)s))
        {% endif %}
        {% if limit %}
        LIMIT {{ limit }}
        {% endif %}
        """
    return db_query(ctx, 'metadb', query, cluster_ids=cluster_ids, limit=limit)


def get_cloud_features(ctx, cloud_id):
    query = """
        SELECT
            array_agg(flag_name) "features"
        FROM dbaas.clouds
        JOIN dbaas.cloud_feature_flags USING (cloud_id)
        WHERE cloud_ext_id = %(cloud_id)s
        """
    result = db_query(ctx, 'metadb', query, cloud_id=cloud_id)

    if result:
        return result[0]['features'] or []
    else:
        return []


def enable_feature_on_cloud(ctx, cloud_id, feature):
    query = """
        INSERT INTO dbaas.cloud_feature_flags (cloud_id, flag_name)
        SELECT
            cl.cloud_id,
            %(feature)s
        FROM dbaas.clouds cl
        WHERE cl.cloud_ext_id = %(cloud_id)s
          AND NOT EXISTS (SELECT 1 FROM dbaas.cloud_feature_flags
                          WHERE cloud_id = cl.cloud_id AND flag_name = %(feature)s )
        """
    db_query(ctx, 'metadb', query, cloud_id=cloud_id, feature=feature, fetch=False)


def disable_feature_on_cloud(ctx, cloud_id, feature):
    query = """
        DELETE FROM dbaas.cloud_feature_flags
        WHERE flag_name = %(feature)s
          AND cloud_id IN (SELECT cloud_id FROM dbaas.clouds
                           WHERE cloud_ext_id = %(cloud_id)s)
        """
    db_query(ctx, 'metadb', query, cloud_id=cloud_id, feature=feature, fetch=False)


def update_cloud_quota(
    ctx,
    cloud_id,
    add_cpu=None,
    add_gpu=None,
    add_memory=None,
    add_disk_space=None,
    disk_quota_type=None,
    add_ssd_space=None,
    add_hdd_space=None,
    add_clusters=None,
):
    """
    Updates cloud quota in metadb.
    """
    add_disk_quotas = DiskQuotas(add_disk_space, disk_quota_type, add_ssd_space, add_hdd_space)

    query = """
        SELECT code.update_cloud_quota(
            i_cloud_ext_id => %(cloud_id)s,
            i_delta        => code.make_quota(
                i_cpu       => coalesce(%(add_cpu)s, 0.0)::real,
                i_gpu       => coalesce(%(add_gpu)s, 0)::bigint,
                i_memory    => coalesce(%(add_memory)s, 0)::bigint,
                i_ssd_space => coalesce(%(add_ssd_space)s, 0)::bigint,
                i_hdd_space => coalesce(%(add_hdd_space)s, 0)::bigint,
                i_clusters  => coalesce(%(add_clusters)s, 0)::bigint
            ),
            i_x_request_id => NULL
        )
        """

    result = db_query(
        ctx,
        'metadb',
        query,
        cloud_id=cloud_id,
        add_cpu=add_cpu,
        add_gpu=add_gpu,
        add_memory=add_memory,
        add_ssd_space=add_disk_quotas.ssd_space,
        add_hdd_space=add_disk_quotas.hdd_space,
        add_clusters=add_clusters,
        fetch_single=True,
    )

    if not result:
        raise ClickException(f'Cloud "{cloud_id}" not found.')


def update_cloud_used_resources(
    ctx,
    cloud_id,
    add_cpu=None,
    add_gpu=None,
    add_memory=None,
    add_disk_space=None,
    disk_quota_type=None,
    add_ssd_space=None,
    add_hdd_space=None,
    add_clusters=None,
):
    """
    Updates cloud's used resources in metadb.
    """
    add_disk_quotas = DiskQuotas(add_disk_space, disk_quota_type, add_ssd_space, add_hdd_space)

    query = """
        SELECT code.update_cloud_usage(
            i_cloud_id   => cloud_id,
            i_delta      => code.make_quota(
                i_cpu       => coalesce(%(add_cpu)s, 0.0)::real,
                i_gpu       => coalesce(%(add_gpu)s, 0)::bigint,
                i_memory    => coalesce(%(add_memory)s, 0)::bigint,
                i_ssd_space => coalesce(%(add_ssd_space)s, 0)::bigint,
                i_hdd_space => coalesce(%(add_hdd_space)s, 0)::bigint,
                i_clusters  => coalesce(%(add_clusters)s, 0)::bigint
            ),
            i_x_request_id => NULL
        )
        FROM dbaas.clouds
        WHERE cloud_ext_id = %(cloud_id)s
        """

    result = db_query(
        ctx,
        'metadb',
        query,
        cloud_id=cloud_id,
        add_cpu=add_cpu,
        add_gpu=add_gpu,
        add_memory=add_memory,
        add_ssd_space=add_disk_quotas.ssd_space,
        add_hdd_space=add_disk_quotas.hdd_space,
        add_clusters=add_clusters,
        fetch_single=True,
    )

    if not result:
        raise ClickException(f'Cloud "{cloud_id}" not found.')


class DiskQuotas:
    def __init__(self, disk_space, disk_type, ssd_space, hdd_space):
        if ssd_space is None:
            ssd_space = 0
        if hdd_space is None:
            hdd_space = 0
        if disk_space:
            if disk_type == 'ssd':
                ssd_space += disk_space
            elif disk_type == 'hdd':
                hdd_space += disk_space
            else:
                raise RuntimeError(f'Unexpected disk quota type: {disk_type}')

        self.ssd_space = ssd_space
        self.hdd_space = hdd_space

File: cloud/mdb/mdb-internal-api/internal/metadb/pg/cloud.go
package pg

import (
	"context"

	"github.com/jackc/pgtype"
	"github.com/jmoiron/sqlx"

	"a.yandex-team.ru/cloud/mdb/internal/sqlutil"
	"a.yandex-team.ru/cloud/mdb/internal/sqlutil/sqlerrors"
	"a.yandex-team.ru/cloud/mdb/mdb-internal-api/internal/metadb"
)

var (
	querySelectCloud = sqlutil.Stmt{
		Name: "SelectCloud",
		Query: `
SELECT
    cloud_id,
    cloud_ext_id,
    cpu_quota,
    gpu_quota,
    memory_quota,
    ssd_space_quota,
    hdd_space_quota,
    clusters_quota,
    memory_used,
    ssd_space_used,
    hdd_space_used,
    cpu_used,
    gpu_used,
    clusters_used,
    code.get_cloud_feature_flags(cloud_id) AS feature_flags
FROM
    code.get_cloud(
        i_cloud_id => :cloud_id,
        i_cloud_ext_id => :cloud_ext_id
    )`,
	}

	queryCreateCloud = sqlutil.Stmt{
		Name: "CreateCloud",
		Query: `
SELECT
    cloud_id,
    cloud_ext_id,
    cpu_quota,
    gpu_quota,
    memory_quota,
    ssd_space_quota,
    hdd_space_quota,
    clusters_quota,
    memory_used,
    ssd_space_used,
    hdd_space_used,
    cpu_used,
    gpu_used,
    clusters_used
  FROM code.add_cloud(
    i_cloud_ext_id => :cloud_ext_id,
    i_quota => code.make_quota(
        i_cpu       => CAST(:cpu_quota AS real),
        i_gpu       => CAST(:gpu_quota AS bigint),
        i_memory    => CAST(:memory_quota AS bigint),
        i_ssd_space => CAST(:ssd_space_quota AS bigint),
        i_hdd_space => CAST(:hdd_space_quota AS bigint),
        i_clusters  => CAST(:clusters_quota AS bigint)
    ),
    i_x_request_id => :x_request_id
)`,
	}

	queryLockCloud = sqlutil.Stmt{
		Name: "LockCloud",
		// language=PostgreSQL
		Query: `
SELECT
    cloud_id,
    cloud_ext_id,
    cpu_quota,
    gpu_quota,
    memory_quota,
    ssd_space_quota,
    hdd_space_quota,
    clusters_quota,
    memory_used,
    ssd_space_used,
    hdd_space_used,
    cpu_used,
    gpu_used,
    clusters_used,
    code.get_cloud_feature_flags(cloud_id) AS feature_flags
FROM code.lock_cloud(:cloud_id)
`,
	}

	querySetCloudQuota = sqlutil.Stmt{
		Name: "SetCloudQuota",
		// language=PostgreSQL
		Query: `
SELECT cloud_id,
       cloud_ext_id,
       cpu_quota,
       gpu_quota,
       memory_quota,
       ssd_space_quota,
       hdd_space_quota,
       clusters_quota,
       memory_used,
       ssd_space_used,
       hdd_space_used,
       cpu_used,
       gpu_used,
       clusters_used
FROM code.set_cloud_quota(
        i_cloud_ext_id => :cloud_ext_id,
        i_quota => code.make_quota(
                i_cpu => CAST(coalesce(:set_cpu, 0.0) AS real),
                i_gpu => CAST(coalesce(:set_gpu, 0) AS bigint),
                i_memory => CAST(coalesce(:set_memory, 0) AS bigint),
                i_ssd_space => CAST(coalesce(:set_ssd_space, 0) AS bigint),
                i_hdd_space => CAST(coalesce(:set_hdd_space, 0) AS bigint),
                i_clusters => CAST(coalesce(:set_clusters, 0) AS bigint)
            ),
        i_x_request_id => :x_request_id
    )
`,
	}

	queryUpdateCloudUsedQuota = sqlutil.Stmt{
		Name: "UpdateCloudUsedQuota",
		// language=PostgreSQL
		Query: `
SELECT
    cloud_id,
    cloud_ext_id,
    cpu_quota,
    gpu_quota,
    memory_quota,
    ssd_space_quota,
    hdd_space_quota,
    clusters_quota,
    memory_used,
    ssd_space_used,
    hdd_space_used,
    cpu_used,
    gpu_used,
    clusters_used
FROM code.update_cloud_usage(
    i_cloud_id   => :cloud_id,
    i_delta      => code.make_quota(
        i_cpu       => CAST(coalesce(:add_cpu, 0.0) AS real),
        i_gpu       => CAST(coalesce(:add_gpu, 0) AS bigint),
        i_memory    => CAST(coalesce(:add_memory, 0) AS bigint),
        i_ssd_space => CAST(coalesce(:add_ssd_space, 0) AS bigint),
        i_hdd_space => CAST(coalesce(:add_hdd_space, 0) AS bigint),
        i_clusters  => CAST(coalesce(:add_clusters, 0) AS bigint)
    ),
    i_x_request_id => :x_request_id
)
`,
	}

	queryGetDefaultFeatureFlags = sqlutil.Stmt{
		Name:  "GetDefaultFeatureFlags",
		Query: `SELECT flag_name FROM dbaas.default_feature_flags`,
	}
)

func (b *Backend) CloudByCloudID(ctx context.Context, cloudID string) (metadb.Cloud, error) {
	return b.queryCloud(ctx, querySelectCloud, map[string]interface{}{"cloud_id": cloudID, "cloud_ext_id": ""})
}

func (b *Backend) CloudByCloudExtID(ctx context.Context, cloudExtID string) (metadb.Cloud, error) {
	return b.queryCloud(ctx, querySelectCloud, map[string]interface{}{"cloud_ext_id": cloudExtID, "cloud_id": nil})
}

func (b *Backend) CreateCloud(ctx context.Context, cextid string, quota metadb.Resources, reqID string) (metadb.Cloud, error) {
	return b.queryCloud(
		ctx,
		queryCreateCloud,
		map[string]interface{}{
			"cloud_ext_id":    cextid,
			"cpu_quota":       quota.CPU,
			"gpu_quota":       quota.GPU,
			"memory_quota":    quota.Memory,
			"ssd_space_quota": quota.SSDSpace,
			"hdd_space_quota": quota.HDDSpace,
			"clusters_quota":  quota.Clusters,
			"x_request_id":    reqID,
		},
	)
}

func (b *Backend) LockCloud(ctx context.Context, cloudID int64) (metadb.Cloud, error) {
	return b.queryCloud(
		ctx,
		queryLockCloud,
		map[string]interface{}{
			"cloud_id": cloudID,
		},
	)
}

func (b *Backend) UpdateCloudQuota(ctx context.Context, cloudExtID string, quota metadb.Resources, reqID string) (metadb.Cloud, error) {
	return b.queryCloud(
		ctx,
		querySetCloudQuota,
		map[string]interface{}{
			"cloud_ext_id":  cloudExtID,
			"set_cpu":       quota.CPU,
			"set_gpu":       quota.GPU,
			"set_memory":    quota.Memory,
			"set_ssd_space": quota.SSDSpace,
			"set_hdd_space": quota.HDDSpace,
			"set_clusters":  quota.Clusters,
			"x_request_id":  reqID,
		},
	)
}

func (b *Backend) UpdateCloudUsedQuota(ctx context.Context, cloudID int64, quota metadb.Resources, reqID string) (metadb.Cloud, error) {
	return b.queryCloud(
		ctx,
		queryUpdateCloudUsedQuota,
		map[string]interface{}{
			"cloud_id":      cloudID,
			"add_cpu":       quota.CPU,
			"add_gpu":       quota.GPU,
			"add_memory":    quota.Memory,
			"add_ssd_space": quota.SSDSpace,
			"add_hdd_space": quota.HDDSpace,
			"add_clusters":  quota.Clusters,
			"x_request_id":  reqID,
		},
	)
}

func (b *Backend) DefaultFeatureFlags(ctx context.Context) ([]string, error) {
	var sourceFeatureFlags []string
	parser := func(rows *sqlx.Rows) error {
		var flag string
		if err := rows.Scan(&flag); err != nil {
			return err
		}

		sourceFeatureFlags = append(sourceFeatureFlags, flag)
		return nil
	}

	_, err := sqlutil.QueryTx(
		ctx,
		queryGetDefaultFeatureFlags,
		map[string]interface{}{},
		parser,
		b.logger,
	)
	if err != nil {
		return nil, err
	}

	return sourceFeatureFlags, nil
}

func (b *Backend) queryCloud(ctx context.Context, query sqlutil.Stmt, args map[string]interface{}) (metadb.Cloud, error) {
	var cloud Cloud
	parser := func(rows *sqlx.Rows) error {
		return rows.StructScan(&cloud)
	}

	count, err := sqlutil.QueryTx(
		ctx,
		query,
		args,
		parser,
		b.logger,
	)
	if err != nil {
		return metadb.Cloud{}, wrapError(err)
	}
	if count == 0 {
		return metadb.Cloud{}, sqlerrors.ErrNotFound
	}

	return cloudFromDB(cloud), nil
}

type Cloud struct {
	CloudID       int64            `db:"cloud_id"`
	CloudExtID    string           `db:"cloud_ext_id"`
	CPUQuota      float64          `db:"cpu_quota"`
	GPUQuota      int64            `db:"gpu_quota"`
	MemoryQuota   int64            `db:"memory_quota"`
	SSDSpaceQuota int64            `db:"ssd_space_quota"`
	HDDSpaceQuota int64            `db:"hdd_space_quota"`
	ClustersQuota int64            `db:"clusters_quota"`
	CPUUsed       float64          `db:"cpu_used"`
	GPUUsed       int64            `db:"gpu_used"`
	MemoryUsed    int64            `db:"memory_used"`
	SSDSpaceUsed  int64            `db:"ssd_space_used"`
	HDDSpaceUsed  int64            `db:"hdd_space_used"`
	ClustersUsed  int64            `db:"clusters_used"`
	FeatureFlags  pgtype.TextArray `db:"feature_flags"`
}

func cloudFromDB(cloud Cloud) metadb.Cloud {
	c := metadb.Cloud{
		CloudID:    cloud.CloudID,
		CloudExtID: cloud.CloudExtID,
		Quota: metadb.Resources{
			CPU:      cloud.CPUQuota,
			GPU:      cloud.GPUQuota,
			Memory:   cloud.MemoryQuota,
			SSDSpace: cloud.SSDSpaceQuota,
			HDDSpace: cloud.HDDSpaceQuota,
			Clusters: cloud.ClustersQuota,
		},
		Used: metadb.Resources{
			CPU:      cloud.CPUUsed,
			GPU:      cloud.GPUUsed,
			Memory:   cloud.MemoryUsed,
			SSDSpace: cloud.SSDSpaceUsed,
			HDDSpace: cloud.HDDSpaceUsed,
			Clusters: cloud.ClustersUsed,
		},
	}

	for _, ff := range cloud.FeatureFlags.Elements {
		c.FeatureFlags = append(c.FeatureFlags, ff.String)
	}

	return c
}
